File: frontend/src/lib/api/contacts.ts
import { api, post, put, del, buildQuery } from './client';

export interface Contact {
	id: number;
	name: string;
	contact_type_id: number | null;
	contact_type_name: string | null;
	farm_number: string | null;
	phone_cell: string | null;
	phone_home: string | null;
	phone_work: string | null;
	email: string | null;
	company_name: string | null;
	description: string | null;
	active: boolean;
	created_at: string;
}

export interface CreateContact {
	name: string;
	type_id?: number;
	farm_number?: string;
	active?: boolean;
	phone_cell?: string;
	phone_home?: string;
	phone_work?: string;
	email?: string;
	company_name?: string;
	description?: string;
}

export interface UpdateContact {
	name?: string;
	type_id?: number;
	farm_number?: string;
	active?: boolean;
	phone_cell?: string;
	phone_home?: string;
	phone_work?: string;
	email?: string;
	company_name?: string;
	description?: string;
}

export interface ContactFilter {
	page?: number;
	per_page?: number;
}

export function listContacts(filter: ContactFilter = {}) {
	return api<{ data: Contact[]; total: number }>(`/contacts${buildQuery(filter)}`);
}

export function createContact(data: CreateContact) {
	return post<{ data: Contact }>('/contacts', data);
}

export function updateContact(id: number, data: UpdateContact) {
	return put<{ data: Contact }>(`/contacts/${id}`, data);
}

export function deleteContact(id: number) {
	return del<{ message: string }>(`/contacts/${id}`);
}

File: backend/src/handlers/contacts.rs
use axum::extract::{Path, Query, State};
use axum::routing::{get, put};
use axum::{Json, Router};
use serde_json::{Value, json};

use crate::errors::AppError;
use crate::middleware::auth::Claims;
use crate::models::contact::{ContactFilter, CreateContact, UpdateContact};
use crate::models::pagination::paginated;
use crate::services::contact_service;
use crate::state::AppState;

pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/contacts", get(list).post(create))
        .route("/contacts/{id}", put(update).delete(remove))
}

#[utoipa::path(
    get,
    path = "/api/v1/contacts",
    responses(
        (status = 200, description = "List of contacts", body = serde_json::Value),
        (status = 401, description = "Unauthorized")
    ),
    params(ContactFilter),
    security(("cookie_auth" = []))
)]
async fn list(
    _claims: Claims,
    State(state): State<AppState>,
    Query(filter): Query<ContactFilter>,
) -> Result<Json<Value>, AppError> {
    let pool = &state.pool;
    let f = &filter;
    paginated(
        filter.page,
        filter.per_page,
        || contact_service::list(pool, f),
        || contact_service::count(pool),
    )
    .await
}

#[utoipa::path(
    post,
    path = "/api/v1/contacts",
    request_body = CreateContact,
    responses(
        (status = 201, description = "Contact created", body = serde_json::Value),
        (status = 400, description = "Validation error"),
        (status = 401, description = "Unauthorized"),
        (status = 403, description = "Admin access required")
    ),
    security(("cookie_auth" = []))
)]
async fn create(
    _admin: crate::middleware::auth::AdminGuard,
    State(state): State<AppState>,
    Json(req): Json<CreateContact>,
) -> Result<Json<Value>, AppError> {
    req.validate()?;
    let item = contact_service::create(&state.pool, &req).await?;
    Ok(Json(json!({ "data": item })))
}

#[utoipa::path(
    put,
    path = "/api/v1/contacts/{id}",
    request_body = UpdateContact,
    responses(
        (status = 200, description = "Contact updated", body = serde_json::Value),
        (status = 404, description = "Not found"),
        (status = 401, description = "Unauthorized"),
        (status = 403, description = "Admin access required")
    ),
    params(("id" = i32, Path, description = "Contact ID")),
    security(("cookie_auth" = []))
)]
async fn update(
    _admin: crate::middleware::auth::AdminGuard,
    State(state): State<AppState>,
    Path(id): Path<i32>,
    Json(req): Json<UpdateContact>,
) -> Result<Json<Value>, AppError> {
    req.validate()?;
    let item = contact_service::update(&state.pool, id, &req).await?;
    Ok(Json(json!({ "data": item })))
}

#[utoipa::path(
    delete,
    path = "/api/v1/contacts/{id}",
    responses(
        (status = 200, description = "Contact deleted", body = serde_json::Value),
        (status = 404, description = "Not found"),
        (status = 401, description = "Unauthorized"),
        (status = 403, description = "Admin access required")
    ),
    params(("id" = i32, Path, description = "Contact ID")),
    security(("cookie_auth" = []))
)]
async fn remove(
    _admin: crate::middleware::auth::AdminGuard,
    State(state): State<AppState>,
    Path(id): Path<i32>,
) -> Result<Json<Value>, AppError> {
    contact_service::delete(&state.pool, id).await?;
    Ok(Json(json!({ "message": "Удалено" })))
}
